Userland/Libraries/LibWeb/HTML/DecodedImageData.h

```diff
@@ -7,38 +7,30 @@
 #pragma once
 
 #include <AK/RefCounted.h>
-#include <LibGfx/Forward.h>
+#include <LibGfx/Size.h>
 #include <LibJS/Heap/Cell.h>
+#include <LibWeb/PixelUnits.h>
 
 namespace Web::HTML {
 
 // https://html.spec.whatwg.org/multipage/images.html#img-req-data
-class DecodedImageData final : public RefCounted<DecodedImageData> {
+class DecodedImageData : public RefCounted<DecodedImageData> {
 public:
-    struct Frame {
-        RefPtr<Gfx::Bitmap const> bitmap;
-        int duration { 0 };
-    };
+    virtual ~DecodedImageData();
 
-    static ErrorOr<NonnullRefPtr<DecodedImageData>> create(Vector<Frame>&&, size_t loop_count, bool animated);
-    ~DecodedImageData();
+    virtual RefPtr<Gfx::Bitmap const> bitmap(size_t frame_index, Gfx::IntSize = {}) const = 0;
+    virtual int frame_duration(size_t frame_index) const = 0;
 
-    RefPtr<Gfx::Bitmap const> bitmap(size_t frame_index) const;
-    int frame_duration(size_t frame_index) const;
+    virtual size_t frame_count() const = 0;
+    virtual size_t loop_count() const = 0;
+    virtual bool is_animated() const = 0;
 
-    size_t frame_count() const { return m_frames.size(); }
-    size_t loop_count() const { return m_loop_count; }
-    bool is_animated() const { return m_animated; }
+    virtual Optional<CSSPixels> intrinsic_width() const = 0;
+    virtual Optional<CSSPixels> intrinsic_height() const = 0;
```

Comment:
I know this is supposed to be a first cut, but I'm not sure if `CSSPixels` is the correct type here.

As far as I have understood it, the `ImageProvider` is the object that bridges the gap between layout engine and actual image data. According to the spec, `CSSPixels` are independent of zoom and device-scaling. On your local machine this probably didn't matter all too much since you have a 1:1 mapping of image pixels to `CSSPixels`, however, this changes in the aforementioned contexts.

Wouldn't it be better to just use "actual image pixels" (i.e. `int`s..?) here and then convert them to their device and context specific `CSSPixels` counter part in the `ImageProvider`?

Reply:
Forget it... I thought it said "... dependent" but it says "... independent" which negates my point entirely.